src/main/java/org/opensearch/knn/index/query/rescore/RescoreSearcher.java

```diff
@@ -0,0 +1,96 @@
+import java.util.stream.Collectors;
+
+@AllArgsConstructor
+public final class RescoreSearcher {
+
+    private final ResultUtil resultUtil;
+
+    /**
+     * Executes a k-NN query by executing a first pass search over an index with quantized vectors with
+     * an oversample factor and then rescoring down to k results for the whole shard.
+     *
+     * @param indexSearcher The index searcher for the current search
+     * @param leafReaderContexts The leaf reader contexts for the current search
+     * @param knnWeight The knn weight for the current search
+     * @return The top k results for the shard
+     * @throws IOException If an error occurs during the search
+     */
+    public TopDocs[] search(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        int firstPassK,
+        int finalK
+    ) throws IOException {
+        List<Map<Integer, Float>> perLeafResults = firstPassSearch(indexSearcher, leafReaderContexts, knnWeight, firstPassK);
+        return rescoreResults(indexSearcher, leafReaderContexts, knnWeight, perLeafResults, finalK);
+    }
+
+    private List<Map<Integer, Float>> firstPassSearch(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        int k
+    ) throws IOException {
+        List<Callable<Map<Integer, Float>>> tasks = new ArrayList<>(leafReaderContexts.size());
+        for (LeafReaderContext leafReaderContext : leafReaderContexts) {
+            tasks.add(() -> firstPassSearchLeaf(leafReaderContext, knnWeight));
+        }
+        return resultUtil.reduceToTopK(indexSearcher.getTaskExecutor().invokeAll(tasks), k);
+    }
+
+    private Map<Integer, Float> firstPassSearchLeaf(LeafReaderContext ctx, KNNWeight queryWeight) throws IOException {
+        Map<Integer, Float> leafDocScores = queryWeight.searchLeaf(ctx);
+        if (leafDocScores.isEmpty()) {
+            return leafDocScores;
+        }
+
+        Bits liveDocs = ctx.reader().getLiveDocs();
+        return leafDocScores.entrySet()
+            .stream()
+            .filter(entry -> liveDocs == null || liveDocs.get(entry.getKey()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private TopDocs[] rescoreResults(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        List<Map<Integer, Float>> perLeafResults,
+        int k
+    ) throws IOException {
+        List<Callable<TopDocs>> topDocTasks = new ArrayList<>(leafReaderContexts.size());
+        for (int i = 0; i < perLeafResults.size(); i++) {
+            BitSet convertedBitSet = resultUtil.resultMapToMatchBitSet(perLeafResults.get(i));
+            LeafReaderContext leafReaderContext = leafReaderContexts.get(i);
+            topDocTasks.add(
+                () -> resultUtil.resultMapToTopDocs(
```

Comment:
this conversion of mapping docs to bitset and then bitset to iterator has potential to incur latency in case of large number segments and high value of first pass of K. 

If you see these bitsets are getting converted to iterators only at the end here: https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/filtered/FilteredIdsKNNIterator.java#L39

So, I see there is a potential to reduce all this extra conversion.

Another idea could be to write our own implementation of DISI which underline uses a Map/List<Integer> for iteration.

Reply:
It seems it would be easiest to switch BitSet to DocIdSetIterator in the FilteredIdsKNNIterator (https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/filtered/FilteredIdsKNNIterator.java#L33) and then create a wrapper of the DocIDSetIterator around the Map of results to avoid the conversion from map to bitset. I can make this change - but before I do, want to get your thoughts?

Reply:
or, I can take for another review - what do you think?

Reply:
Lets do it in separate review. But it should be done before the release of disk based vector search feature.